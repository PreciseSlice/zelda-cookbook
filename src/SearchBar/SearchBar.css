.SearchBar {
  display: grid;
  grid-template-columns: auto auto auto;
  align-items: center;
  margin: 20px auto;
  width: fit-content; }

.search-bar {
  font-size: 1.5em;
  width: 300px;
  border: 1px solid black;
  height: 33px;
  padding-left: 5px; }

.search-bar__ingredient {
  display: inline-block;
  position: relative; }

.search-bar--hidden {
  display: none; }

.search-button {
  width: 130px;
  border: 1px solid black;
  text-align: center;
  font-size: 1.5em;
  background-color: lightgray;
  color: gray; }

.search-button--selected {
  background-color: white;
  color: black; }

.submit-button {
  font-size: 1.5em;
  border: 1px solid black; }

.search-results__container {
  width: 300px;
  border: 1px solid black;
  display: none;
  position: absolute; }

.search-results__containter--show {
  display: block; }

.search-results__item {
  margin: 0;
  padding: 20px; }
  .search-results__item:hover {
    background-color: gray;
    font-weight: bold; }
