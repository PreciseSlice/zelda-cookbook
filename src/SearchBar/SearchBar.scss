.search-bar {
  font-size: 1.5em;
  width: 300px;
  border: 1px solid black;
  height: 33px;
  padding-left: 5px;
}

.search-form {
  display: inline-block;
}

.search-bar--hidden {
  display: none;
}

.search-button {
  width: 130px;
  border: 1px solid black;
  text-align: center;
  font-size: 1.5em;
  // padding: 10px;
  background-color: lightgray;
  color: gray;
}

.search-button--selected {
  background-color: white;
  color: black
}

.submit-button {
  font-size: 1.5em;
  border: 1px solid black;
}

.search-results__container {
  width: 300px;
  border: 1px solid black;
  margin-left: 260px;
  display: none;
}

.search-results__containter--show {
  display: block;
}

.search-results__item {
  margin: 0;
  padding: 20px;

  &:hover {
    background-color: gray;
    font-weight: bold;
  }
}