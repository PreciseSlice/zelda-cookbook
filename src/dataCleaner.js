export const getIngredientNames = rawData => {
  return rawData.map(({name, id}) => ({name, id}))
}

// export const getRecipeData = rawData => {
//   return rawData.map(({name, id}) => ({name, id}))
// }
