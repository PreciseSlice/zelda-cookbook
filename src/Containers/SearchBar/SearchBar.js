import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { getIngredientNames } from '../../dataCleaner';
import IngredientTag from '../../Components/IngredientTag/IngredientTag';
import './SearchBar.css';
import {
  addIngredientFilter,
  removeIngredientFilter,
  updateNameFilter
} from '../../actions';

// this seems real gross, and i'd like to change it, but it works for now.
let timeout;

export class SearchBar extends Component {
  constructor(props) {
    super(props);
    this.state = {
      ingredientSearch: '',
      recipeSearch: '',
      ingredientButton: true,
      recipeButton: true,
      ingredientBar: true,
      recipeBar: true,
      searchFocus: false,
    };
  }

  handleChange = event => {
    const { name, value } = event.target;
    this.setState({ [name]: value });
    if (name === 'recipeSearch') {
      clearTimeout(timeout);
      timeout = setTimeout(() => {
        this.props.updateNameFilter(value);
      }, 500);
    }
  }

  handleClick = event => {
    const { id } = event.target;
    this.toggleSearchBar(id);
    this.toggleSearchButton(id);
  }

  toggleSearchBar = id => {
    const { ingredientBar, recipeBar } = this.state;
    if (this.state[`${id}Bar`] === false) {
      this.setState({
        ingredientBar: !ingredientBar,
        recipeBar: !recipeBar,
        ingredientSearch: '',
        recipeSearch: ''
      });
    }
  }

  toggleSearchButton = id => {
    const { ingredientButton, recipeButton } = this.state;
    if (this.state[`${id}Button`] === false) {
      this.setState({
        ingredientButton: !ingredientButton,
        recipeButton: !recipeButton
      });
    }
  }

  ingredientListToRender = () => {
    const ingredientNames = getIngredientNames(this.props.ingredients);
    return ingredientNames
      .filter(food => food.name.toLowerCase().includes(
        this.state.ingredientSearch.toLowerCase())
      )
      .map(food => <p
        key={food.id} 
        className='search-results__item'
        id={food.id}
        onClick={this.selectIngredient}
      >
        {food.name}
      </p>);
  }

  selectIngredient = (event) => {
    const { ingredientFilter } = this.props;
    const { innerText , id } = event.target;
    const newIngredient = { id, name: innerText };
    const checkDupe = ingredientFilter.some(tag => tag.name === newIngredient.name);
    if ( !checkDupe && ingredientFilter.length < 5 ) {
      this.props.addIngredientFilter(newIngredient);
    }
    this.setState({ searchFocus: false, ingredientSearch: '' });
  }

  handleFocus = () => {
    this.setState({ searchFocus: true });
  }

  handleBlur = () => {
    setTimeout(() => {
      this.setState({ searchFocus: false });
    }, 150);
  }

  tagsToRender = () => {
    const { ingredientFilter } = this.props;
    return ingredientFilter.map(tag => <IngredientTag
      {...tag}
      key={tag.id}
      remove={this.removeIngredient}
    />);
  }

  removeIngredient = (event) => {
    const { id } = event.target;
    this.props.removeIngredientFilter(id);
  }
  
  render() {
    const {
      ingredientBar,
      recipeBar,
      ingredientButton,
      recipeButton
    } = this.state;

    return (
      <div>
        <section className="SearchBar">
          <button
            className={`search-button
              ${ingredientButton ? 'search-button--selected' : ''}
            `}
            onClick={this.handleClick}
            id="ingredient"
          >
            Ingredient
          </button>
          <button
            className={`search-button
              ${recipeButton ? 'search-button--selected' : ''}
            `}
            onClick={this.handleClick}
            id="recipe"
          >
          Recipe
          </button>
          <input
            type="text"
            placeholder="Search recipes"
            value={this.state.recipeSearch}
            name="recipeSearch"
            className={`search-bar
              search-bar__recipe
              ${recipeBar ? '' : 'search-bar--hidden'}
            `}
            onChange={this.handleChange}
          />
          <div 
            className={`
              search-bar__ingredient
              ${ingredientBar ? '' : 'search-bar--hidden'}
            `}
            onFocus={this.handleFocus}
            onBlur={this.handleBlur}
          >
            <input
              type="text"
              placeholder="Select up to 5 ingredients"
              value={this.state.ingredientSearch}
              name="ingredientSearch"
              className={`search-bar
              `}
              onChange={this.handleChange}
            />
            <section className={`
              search-results__container
              ${this.state.searchFocus ? 'search-results__containter--show' : ''}
            `}>
              {this.ingredientListToRender()}
            </section>
          </div>
        </section>
        <section className="ingredientTag-container">
          {this.tagsToRender()}
        </section>
      </div>
    );
  }
}

const { arrayOf, shape, string, number, array, func } = PropTypes;

SearchBar.propTypes = {
  ingredients: arrayOf(shape({
    name: string,
    id: number
  })),
  ingredientFilter: array,
  addIngredientFilter: func,
  removeIngredientFilter: func,
  updateNameFilter: func
};

export const MSTP = ({ingredients, ingredientFilter}) => ({
  ingredients,
  ingredientFilter
});

export const MDTP = dispatch => ({
  addIngredientFilter: ingredient => dispatch(addIngredientFilter(ingredient)),
  removeIngredientFilter: id => dispatch(removeIngredientFilter(id)),
  updateNameFilter: name => dispatch(updateNameFilter(name))
});

export default connect(MSTP, MDTP)(SearchBar);
