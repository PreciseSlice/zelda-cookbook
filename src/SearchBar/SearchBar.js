import React, { Component } from 'react';
import { getIngredients } from '../apiCalls';
import { getIngredientNames } from '../dataCleaner';
import IngredientTag from '../IngredientTag/IngredientTag';
import './SearchBar.css';

export default class SearchBar extends Component {
  constructor(props) {
    super(props);
    this.state = {
      ingredients: [],
      recipes: [],
      ingredientSearch: '',
      recipeSearch: '',
      ingredientButton: false,
      recipeButton: true,
      ingredientBar: false,
      recipeBar: true,
      ingredientTags: [],
      searchFocus: false,
    };
  }

  componentDidMount = async () => {
    const rawIngredients = await getIngredients();
    const ingredients = getIngredientNames(rawIngredients);
    this.setState({ ingredients });
  }

  handleChange = event => {
    const { name, value } = event.target;
    this.setState({ [name]: value });
  }

  handleClick = event => {
    const { id } = event.target;
    this.toggleSearchBar(id);
    this.toggleSearchButton(id);
  }

  toggleSearchBar = id => {
    const { ingredientBar, recipeBar } = this.state
    if (this.state[`${id}Bar`] === false) {
      this.setState({
        ingredientBar: !ingredientBar,
        recipeBar: !recipeBar
      })
    }
  }

  toggleSearchButton = id => {
    const { ingredientButton, recipeButton } = this.state
    if (this.state[`${id}Button`] === false) {
      this.setState({
        ingredientButton: !ingredientButton,
        recipeButton: !recipeButton
      })
    }
  }

  ingredientListToRender = () => {
    return this.state.ingredients
      .filter(food => food.name.toLowerCase().includes(
        this.state.ingredientSearch.toLowerCase())
      )
      .map(food => <p
        key={food.id} 
        className='search-results__item'
        id={food.id}
        onClick={this.selectIngredient}
      >
        {food.name}
      </p>)
  }

  selectIngredient = (event) => {
    const { ingredientTags } = this.state
    const { innerText , id } = event.target;
    const newIngredient = { id, name: innerText };
    const checkDupe = ingredientTags.some(tag => tag.name === newIngredient.name)
    if ( !checkDupe && ingredientTags.length < 5 ) {
      this.setState({ ingredientTags: [...ingredientTags, newIngredient]})
    }
  }

  handleFocus = () => {
    this.setState({ searchFocus: true })
  }

  handleBlur = () => {
    setTimeout(() => {
      this.setState({ searchFocus: false })
    }, 150)
  }

  tagsToRender = () => {
    const { ingredientTags } = this.state;
    return ingredientTags.map(tag => <IngredientTag
      {...tag}
      key={tag.id}
      remove={this.removeIngredient}
    />)
  }

  removeIngredient = (event) => {
    const { id } = event.target;
    const ingredientTags = this.state.ingredientTags.filter(tag => tag.id !== id);
    this.setState({ ingredientTags });
  }
  
  render() {
    const {
      ingredientBar,
      recipeBar,
      ingredientButton,
      recipeButton
    } = this.state;

    return (
      <div>
        <section className="SearchBar">
          <button
            className={`search-button
              ${ingredientButton ? 'search-button--selected' : ''}
            `}
            onClick={this.handleClick}
            id="ingredient"
          >
            Ingredient
          </button>
          <button
            className={`search-button
              ${recipeButton ? 'search-button--selected' : ''}
            `}
            onClick={this.handleClick}
            id="recipe"
          >
            Recipe
          </button>
          <input
            type="text"
            placeholder="Search recipes"
            value={this.state.recipeSearch}
            name="recipeSearch"
            className={`search-bar
              search-bar__recipe
              ${recipeBar ? '' : 'search-bar--hidden'}
            `}
            onChange={this.handleChange}
          />
          <div 
            className={`
              search-bar__ingredient
              ${ingredientBar ? '' : 'search-bar--hidden'}
            `}
            onFocus={this.handleFocus}
            onBlur={this.handleBlur}
          >
            <input
              type="text"
              placeholder="Select up to 5 ingredients"
              value={this.state.ingredientSearch}
              name="ingredientSearch"
              className={`search-bar
              `}
              onChange={this.handleChange}
            />
            <section
              className={`search-results__container
                ${this.state.searchFocus ?
                  'search-results__containter--show' :
                  ''
                }
              `}
            >
              {this.ingredientListToRender()}
            </section>
          </div>
        </section>
        <section className="ingredientTag-container">
          {this.tagsToRender()}
        </section>
      </div>
    );
  }
}
