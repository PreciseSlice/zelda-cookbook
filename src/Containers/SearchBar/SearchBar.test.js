/* eslint-disable */
import React from 'react';
import { shallow } from 'enzyme';
import { SearchBar, MSTP, MDTP } from './SearchBar';
import * as dataCleaner from '../../dataCleaner';
import {
  addIngredientFilter,
  removeIngredientFilter,
  updateNameFilter
} from '../../actions';

jest.mock('../../dataCleaner')
jest.useFakeTimers();

describe('SearchBar', () => {
  let wrapper;
  let mockIngredients = []
  let mockIngredientFilter = []
  let mockAddIngredientFilter = jest.fn()
  let mockRemoveIngredientFilter = jest.fn()
  let mockUpdateNameFilter = jest.fn()
  
  beforeEach(() => {
    wrapper = shallow(<SearchBar
      ingredients={ mockIngredients }
      ingredientFilter = { mockIngredientFilter }
      addIngredientFilter = { mockAddIngredientFilter }
      removeIngredientFilter = { mockRemoveIngredientFilter }
      updateNameFilter = { mockUpdateNameFilter }
    />);
  });

  it('should have a default state', () => {
    const expected = {
      ingredientSearch: '',
      recipeSearch: '',
      ingredientButton: false,
      recipeButton: true,
      ingredientBar: false,
      recipeBar: true,
      searchFocus: false,
    };

    expect(wrapper.state()).toEqual(expected);
  });

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('handleChange', () => {
    it('should set state based on the input that is changed', () => {
      const event = { target: { 
        name: 'recipeSearch',
        value: 'asdf'
      }};
      const event2 = { target: { 
        name: 'ingredientSearch',
        value: 'qwerty'
      }};

      wrapper.instance().handleChange(event);
      expect(wrapper.state().recipeSearch).toEqual('asdf');
      wrapper.instance().handleChange(event2);
      expect(wrapper.state().ingredientSearch).toEqual('qwerty');
    });

    it('should wait until the user is finished typing, then call updateNameFilter with the value in recipeSearch', () => {
      const event = { target: { 
        name: 'recipeSearch',
        value: 'Pie'
      }};
      wrapper.instance().handleChange(event)
      jest.runAllTimers()
      expect(mockUpdateNameFilter).toHaveBeenCalledWith('Pie')
    })
  });

  describe('handleClick', () => {
    it('should call toggleSearchBar and toggleSearchButton with the target id', () => {
      const event = { target: {
        id: 'ingredient' 
      }};

      wrapper.instance().toggleSearchBar = jest.fn();
      wrapper.instance().toggleSearchButton = jest.fn();


      wrapper.instance().handleClick(event);
      expect(wrapper.instance().toggleSearchBar).toHaveBeenCalledWith('ingredient');
      expect(wrapper.instance().toggleSearchButton).toHaveBeenCalledWith('ingredient');
    });
  });

  describe('toggleSearchBar', () => {
    it('should match the snapshot when it is called', () => {
      expect(wrapper.instance().toggleSearchBar('ingredient')).toMatchSnapshot();
      expect(wrapper.instance().toggleSearchBar('recipe')).toMatchSnapshot();
    });
  });

  describe('toggleSearchButton', () => {
    it('should match the snapshot when it is called', () => {
      expect(wrapper.instance().toggleSearchBar('ingredient')).toMatchSnapshot();
      expect(wrapper.instance().toggleSearchBar('recipe')).toMatchSnapshot();    });
  });

  describe('ingredientListToRender', () => {
    it('should render a list of ingredients and match the snapshot', () => {
      expect(wrapper.instance().ingredientListToRender()).toMatchSnapshot()
    });

    it('should filter its results based on what the user is searching for', () => {
      wrapper.state().ingredientSearch = 'asdf'
      expect(wrapper.instance().ingredientListToRender()).toMatchSnapshot()
    });
  });

  describe('selectIngredient', () => {
    it('should not call addIngredientFilter if the object is already selected', () => {
      wrapper = shallow(<SearchBar
        ingredients={ mockIngredients }
        ingredientFilter = { [{id: '1', name: 'Apple'}] }
        addIngredientFilter = { mockAddIngredientFilter }
        removeIngredientFilter = { mockRemoveIngredientFilter }
        updateNameFilter = { mockUpdateNameFilter }
      />)

      const mockEvent = { target: {id: '1', innerText: 'Apple'}}
      wrapper.instance().selectIngredient(mockEvent)
      expect(mockAddIngredientFilter).not.toHaveBeenCalled()
    });

    it('should not call addIngredientFilter if there are 5 ingredients already', () => {
      wrapper = shallow(<SearchBar
        ingredients={ mockIngredients }
        ingredientFilter = {[
          {id: '0', name: 'Chicken'},
          {id: '2', name: 'Grape'},
          {id: '3', name: 'Pear'},
          {id: '4', name: 'Fish'},
          {id: '5', name: 'Tuna'}
        ]}
        addIngredientFilter = { mockAddIngredientFilter }
        removeIngredientFilter = { mockRemoveIngredientFilter }
        updateNameFilter = { mockUpdateNameFilter }
      />)
      const mockEvent = { 
        target: {
          id: '1', innerText: 'Apple'
        }
      }
      wrapper.instance().selectIngredient(mockEvent)
      expect(mockAddIngredientFilter).not.toHaveBeenCalled()
    });

    it('should call this.props.addIngredientFilter with a new ingredient', () => {
      const mockEvent = { target: {id: '1', innerText: 'Apple'}}
      const expected = {id: '1', name: 'Apple'}
      wrapper.instance().selectIngredient(mockEvent)
      expect(mockAddIngredientFilter).toHaveBeenCalledWith(expected)
    });
  });

  describe('handleFocus', () => {
    it('should set searchFocus to true', () => {
      wrapper.instance().handleFocus()
      expect(wrapper.state().searchFocus).toEqual(true)
    });
  });

  describe('handleBlur', () => {
    it('should set searchFocus to false', () => {
      wrapper.state().searchFocus = true
      wrapper.instance().handleBlur()
      jest.runAllTimers()
      expect(wrapper.state().searchFocus).toEqual(false)
    });
  });

  describe('tagsToRender', () => {
    it('should return tags depending on what is in the filters ingredients array', () => {
      wrapper = shallow(<SearchBar
        ingredients={ mockIngredients }
        ingredientFilter = {[
          {id: '0', name: 'Chicken'},
          {id: '2', name: 'Grape'},
          {id: '3', name: 'Pear'},
          {id: '4', name: 'Fish'},
          {id: '5', name: 'Tuna'}
        ]}
        addIngredientFilter = { mockAddIngredientFilter }
        removeIngredientFilter = { mockRemoveIngredientFilter }
        updateNameFilter = { mockUpdateNameFilter }
      />)
      expect(wrapper.instance().tagsToRender()).toMatchSnapshot()
    });
  });

  describe('removeIngredient', () => {
    it('should call removeIngredientFilter with the selected id', () => {
      const mockEvent = {target: {id: 1}}

      wrapper.instance().removeIngredient(mockEvent)
      expect(mockRemoveIngredientFilter).toHaveBeenCalledWith(1)
    });
  });

  describe('MSTP', () => {
    it('should return an object with with ingredients and a filter object', () => {
      const mockState = {
        ingredients: [{
          id: '1',
          name: 'Apple',
        }],
        ingredientFilter: []
      }
      const mapped = MSTP(mockState);

      expect(mapped.ingredients).toEqual([{
        id: '1',
        name: 'Apple',
      }]);
      expect(mapped.ingredientFilter).toEqual([]);
    });
  });

  describe('MDTP', () => {
    it('should call dispatch when addIngredientFilter or removeIngredientFilter are called', () => {
      const mockDispatch = jest.fn();
      const mockIngredient = { id: '1', name: 'Apple' };
      const mockId = 1;
      const result = MDTP(mockDispatch);
      const mockName = 'Pie'

      result.addIngredientFilter(mockIngredient);
      expect(mockDispatch).toHaveBeenCalledWith(addIngredientFilter(mockIngredient));

      result.removeIngredientFilter(mockId);
      expect(mockDispatch).toHaveBeenCalledWith(removeIngredientFilter(mockId));

      result.updateNameFilter(mockName);
      expect(mockDispatch).toHaveBeenCalledWith(updateNameFilter(mockName));
    });
  });
});