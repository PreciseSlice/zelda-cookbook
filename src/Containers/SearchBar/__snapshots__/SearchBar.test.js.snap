// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar ingredientListToRender should filter its results based on what the user is searching for 1`] = `Array []`;

exports[`SearchBar ingredientListToRender should render a list of ingredients and match the snapshot 1`] = `
Array [
  <p
    className="search-results__item"
    id={1}
    onClick={[Function]}
>
    Apple
</p>,
]
`;

exports[`SearchBar should match the snapshot 1`] = `
<div>
  <section
    className="SearchBar"
  >
    <button
      className="search-button
                  
                "
      id="ingredient"
      onClick={[Function]}
    >
      Ingredient
    </button>
    <button
      className="search-button
                  search-button--selected
                "
      id="recipe"
      onClick={[Function]}
    >
      Recipe
    </button>
    <input
      className="search-bar
                  search-bar__recipe
                  
                "
      name="recipeSearch"
      onChange={[Function]}
      placeholder="Search recipes"
      type="text"
      value=""
    />
    <div
      className="
                  search-bar__ingredient
                  search-bar--hidden
                "
      onBlur={[Function]}
      onFocus={[Function]}
    >
      <input
        className="search-bar
                    "
        name="ingredientSearch"
        onChange={[Function]}
        placeholder="Select up to 5 ingredients"
        type="text"
        value=""
      />
      <section
        className="
                    search-results__container
                    
                  "
      >
        <p
          className="search-results__item"
          id={1}
          key="1"
          onClick={[Function]}
        >
          Apple
        </p>
      </section>
    </div>
  </section>
  <section
    className="ingredientTag-container"
  />
</div>
`;

exports[`SearchBar tagsToRender should return tags depending on what is in the filters ingredients array 1`] = `
Array [
  <IngredientTag
    id="0"
    name="Chicken"
    remove={[Function]}
/>,
  <IngredientTag
    id="2"
    name="Grape"
    remove={[Function]}
/>,
  <IngredientTag
    id="3"
    name="Pear"
    remove={[Function]}
/>,
  <IngredientTag
    id="4"
    name="Fish"
    remove={[Function]}
/>,
  <IngredientTag
    id="5"
    name="Tuna"
    remove={[Function]}
/>,
]
`;

exports[`SearchBar toggleSearchBar should match the snapshot when it is called 1`] = `undefined`;

exports[`SearchBar toggleSearchBar should match the snapshot when it is called 2`] = `undefined`;

exports[`SearchBar toggleSearchButton should match the snapshot when it is called 1`] = `undefined`;

exports[`SearchBar toggleSearchButton should match the snapshot when it is called 2`] = `undefined`;
